term/src/terminalstate.rs

```diff
@@ -1949,10 +1963,26 @@ impl TerminalState {
                 // We always output at our "best" rate
             }
 
-            Mode::SetDecPrivateMode(DecPrivateMode::Code(DecPrivateModeCode::ReverseVideo))
-            | Mode::ResetDecPrivateMode(DecPrivateMode::Code(DecPrivateModeCode::ReverseVideo)) => {
-                // I'm mostly intentionally ignoring this in favor
-                // of respecting the configured colors
+            Mode::SetDecPrivateMode(DecPrivateMode::Code(DecPrivateModeCode::ReverseVideo)) => {
+                // Turn on reverse video for all of the lines on the
+                // display.
+                self.screen.reverse_video_mode = true;
+                for y in 0..self.screen.physical_rows as i64 {
```

Comment:
In rust, array dimension accesses are bounds-checked by default which means that there is some (small) overhead for C style iteration.  It's preferable to iterate containers directly so that the bounds check happens just once and the individual steps are essentially equivalent to bumping a pointer with no other overhead:

```rust
for line in &mut self.screen.lines {
   line.set_reverse(true);
}
```

Reply:
This is `O(scrollback_size)` whenever the reverse video mode is changed, and some people love having really huge scrollback.  Is reverse video really a property of the line?  Does it make sense to expose `TerminalState::get_reverse_video()` as an accessor to avoid this update pass?

For rendering purposes, we could still keep `LineBits::REVERSE` but teach `terminal_get_lines` to set it when retrieving the lines that need to be rendered:

https://github.com/wez/wezterm/blob/main/mux/src/renderable.rs#L72-L93

```rust
/// Implements Pane::get_lines for Terminal
pub fn terminal_get_lines(
    term: &mut Terminal,
    lines: Range<StableRowIndex>,
) -> (StableRowIndex, Vec<Line>) {
    let screen = term.screen_mut();
    let reverse = term.get_reverse_video(); // added
    let phys_range = screen.stable_range(&lines);
    (
        screen.phys_to_stable_row_index(phys_range.start),
        screen
            .lines
            .iter_mut()
            .skip(phys_range.start)
            .take(phys_range.end - phys_range.start)
            .map(|line| {
                let cloned = line.clone();
                line.clear_dirty();
                cloned.set_reverse(reverse);  // added
                cloned
            })
            .collect(),
    )
}
```

If something like that makes sense, then `terminalstate.rs` would only need to track whether the mode is on or not, and not need to set it on the lines themselves, or deal with that when rewrapping.

Reply:
A few thoughts:

* Setting/clearing reverse ought to only impact lines on the visible screen, not every line in the scrollback.  If the entire scrollback is begin affected (or scanned), then I need to fix that.
* I have the line-level flag there so that reversed lines that enter the scrollback retain the reverse flag.  But that's really just an opinion: the real VT100 had no scrollback, and reverse was only ever a whole-screen setting.  I should see what xterm does: whether it keeps the reverse on for lines that go into scrollback or not.
* The flag on the screen is only so that new lines are also set to reverse.  But if we decide that reverse goes away for lines that enter scrollback, then we can eliminate the line-level flag entirely and just set a reverse flag at physical screen render time.

Reply:
A small bit of exploring shows that for xterm, konsole, and VTE, reverse video is a property of the screen only, and furthermore when set all lines rendered, including those in scrollback, are reversed.

I will remove the line-level attribute and leave it on the screen only.

Reply:
OK, reverse is nicer, thank you!  However, there is one gap remaining: on lines that do not stretch to the full physical pixels, we do not see the reversed background.  Example:

![reverse_incomplete](https://user-images.githubusercontent.com/4357501/125082094-6802f680-e08c-11eb-9a03-402fd192257c.png)

I need to add a bit after render.rs:933, so that the selection and cursor are rendered, and then if there is more physical width remaining fill it with bg_color.  Any suggestions?